### DISCUSSION TITLE
Question about the INDYWV audio format

### DISCUSSION THREAD

**Discussion by justedni** [smlu/ProjectMarduk#9] [General]
Hello!
First, thank you so much for all your hard work with this incredible collection of tools.
I was investigating the INDYWV file format on another game (Star Wars Episode 1 The Phantom Menace), and found your github. Your work on reverse-engineering the header and decompression algorithms is really impressive! I saw in [indywv.h](https://github.com/smlu/Urgon/blob/master/libraries/libim/content/audio/impl/serialization/indywv.h) that there was a "TODO: Implement ADPCM decompression method". Sadly I believe that Phantom Menace doesn't use WVSM compression, the magic numbers don't match, unknown1 is always a positive number. Also, your algorithm only covers stereo files, but in my case, I'm trying to decompress dialog files so they are likely to be mono.
I was just wondering if you could give me some pointers on how you managed to figure out how the decompression worked for Infernal Machine and maybe I can try and implement the rest of the algorithm.
Thanks in advance for your time!

**Comment by smlu** [comment]
Hey!
My starting point was here: [https://web.archive.org/web/20160404215239/http://wiki.xentax.com/index.php?title=Indiana_Jones_And_The_Infernal_Machine_CND](https://web.archive.org/web/20160404215239/http://wiki.xentax.com/index.php?title=Indiana_Jones_And_The_Infernal_Machine_CND). After I learnt IndyWV decompression algorithm I jumped into RE and try to decompiled the rest of the code there. Unfortunately I wasn't able to decompile the other decompression algorithm (no game sound is compressed in that format) which I believe it's some variation of [ADPCM compression](https://en.wikipedia.org/wiki/Adaptive_differential_pulse-code_modulation) algorithm. 
Fyi this is the function signature of the other decompression method: `void __cdecl AudioLib_Uncompress(tAudioCompressorState *pCompressorState, BYTE *pOutData, const BYTE *pInData, int size, unsigned int numChannels, int bStateInited)`. `AudioLib` module is part of LEC's code. The `tAudioCompressorState ` type is `VWCompressorState` struct in my code.

I hope this info helps you in any way!

**Reply by justedni** [reply]
It does help a ton! Thank you so much for your kind reply! I'm going to try the same approach with Phantom Menace until I can get the decompiled code for the decompression algorithm.
Are there any specific tools or resources that you'd recommend for this? I am a programmer but I don't really do a lot of reverse engineering ;)

**Reply by smlu** [reply]
I think the best opensource RE tool would be [ghidra](https://github.com/NationalSecurityAgency/ghidra). It's great because you have assembly code and decompiled pseudo C code side by side, from what I've tried other opensource RE tools this one gives you the best C pseudo code output. It has some learning curve tho, but if you're familiar with C it should expedite the learning process. Besides the RE tools, some AI chat tools like chatgtp, claude, mistral ai etc... can give you some pointers, especially in the fields you're not familiar with, but don't take them for granted 'cause they can be wrong!

**Reply by justedni** [reply]
This is really helpful, thank you so much! I'll see where this leads me. I'm quite familiar with C so I'm not too worried about the learning curve :) I'll make sure to let you know if I manage to crack the case regarding that ADPCM algorithm!

**Reply by justedni** [reply]
I've done it!! I've managed to reverse-engineer the algorithm. I've used Ghidra and also x64dbg for debugging. You were right, it's a bit of a learning curve but this was an amazing opportunity for me to learn how to use those tools and get more familiar with assembly instructions. Locating the algorithm was easy, but then extracting the raw data and making the algorithm work took me several days.
I've published everything on my own github: https://github.com/justedni/indywvtowav .
I took the liberty to add your WVSM implementation as well, this way my tool can work as a "complete" standalone IndyWV to Wav converter. I hope you don't mind (if that's the case please let me know). I made sure to credit you in the readme file.
Thank you so much for taking the time to share your approach with me. This was truly motivating! :)

**Reply by smlu** [reply]
Wow nice, that was fast! Is the decompression algorithm actually variation of ADPCM? 

Btw awesome that you made another tool for this sound format, I'll definitely check it out. Also, it's totally fine with me if you used my implementation for the WASM decompression algorithm.

**Reply by justedni** [reply]
Thanks!! To be honest, I'm not very familiar with ADPCM decompression. I looked it up and found [this implementation on github](https://github.com/Soreing/audio-adpcm-cpp/blob/main/src/adpcm.cpp), and the first thing I noticed is that some of the static data that I extracted from Ghidra (the stepsize table) looks exactly the same. Some of the concepts in the algorithm look very similar. But the INDYWV implementation is way more complex and it has more static data and coefficients. I have no idea what they are for or how they work. In my [implementation ](https://github.com/justedni/indywvtowav/blob/main/indywv.cpp) I kept a lot of the logic as it appeared in Ghidra's pseudo-code. As long as it works!

**Reply by smlu** [reply]
Ah smart move man :). Since I don't have any sample of an audio that is encoded in this format it didn't occurred to just look for any existing solutions. Do you maybe know any other game that uses this audio format?

Btw for your tool do you also plan to support compressing wav to indywv format?
Fyi for another project I have for this infernal machine game, I just copied the RE code for IndyWV compression format and it works.

**Reply by justedni** [reply]
Well you didn't need ADPCM decompression for your Infernal Machine tools anyway so there was no reason to spend time on it! :)
I'd love to know if other games use that format, but the list of LucasArts games is so long, it would take me a while to check them one by one. I believe that LucasArts games used a lot of different audio formats and engines over the years. I don't even fully understand how the INDYWV format can be in both Phantom Menace (which was developed by Big Ape Productions) and Infernal Machine (developed internally by LucasArts). And why does Phantom Menace use raw WAV (for SFX), IMC (for Music) and IndyWV (for dialogs), while Infernal Machine use IndyWV for everything? Really curious!

I didn't think about supporting wav to indywv conversions, because when I looked in Ghidra I couldn't find anything I could use. I didn't think they would have kept the code for that in the game? (usually compression would be an editor-only feature and removed from the game build) But are you saying that you found it when you reverse-engineered Infernal Machine? I'll have to check it out then!

**Reply by smlu** [reply]
>Well you didn't need ADPCM decompression for your Infernal Machine tools anyway so there was no reason to spend time on it! :)

I guess so :) ... tho I'm also working on another project that involves using parts of decompiled code, including this ADPCM compression/decompression code. While I could skip it since it's not really used, I was planning to incorporate it anyway, just to have complete code that mimics the original.

>...I believe that LucasArts games used a lot of different audio formats and engines over the years..

That's right, although most older games (older than games based on sith engine) all incorporated iMUSE system in one way or another. My guess is that parts of that system was then inherited into newer game engines.

> I don't even fully understand how the INDYWV format can be in both Phantom Menace (which was developed by Big Ape Productions) and Infernal Machine (developed internally by LucasArts)

According to this site [https://starwars.fandom.com/wiki/Sith_Engine](https://starwars.fandom.com/wiki/Sith_Engine) the SWE1TPM game uses parts of sith engine, so that might explain why INDYWV audio format is there.

> And why does Phantom Menace use raw WAV (for SFX), IMC (for Music) and IndyWV (for dialogs)

My wild guess is that different audio compression formats were chosen based on a trade-off between quality and compression/decompression ratio. They IndyWV format, in my opinion provides efficient decompression rates but sacrifices audio quality (considering the technology available at that time). It might have been perfect for voice dialogues where high-quality audio wasn't a priority, but prompt availability and uninterrupted flow were crucial. I'm not familiar with the IMC format, but I would assume it offers better sound quality. As for the use of raw WAV format for SFX, it's possible they prioritized original audio quality since the size of each file in RAM was likely negligible compared to voice and music. (or they used low bit rate to save space, and that's sort of compression- not really familiar with the game, just a wild guessing here :) 

> But are you saying that you found it when you reverse-engineered Infernal Machine?

Yes :) 
The reason for this is that the game engine (Jones3D engine aka modified sith engine) supports multiple audio formats:  FORM -AIFFCOMM, WV & WAV. The WAV and AIFFCOMM audio formats are considered uncompressed and are compressed on the fly when audio is loaded to either the WV format (IndyWV) or this ADPCM format based on the number of audio channels: if there are two or less channels, it's compressed as the WV format; otherwise, it's compressed using ADPCM.

**Reply by justedni** [reply]
> the SWE1TPM game uses parts of sith engine

This is really fascinating to me! Thank you for sharing your knowledge with me, you seem quite familiar with the Sith engine in general :)

> My wild guess is that different audio compression formats were chosen based on a trade-off between quality and compression/decompression ratio. 

It would make total sense, I agree with your deductions! Also, I recall there were some nice music transitions in SWE1TPM so that's probably why that game was using the iMuse system.

> The reason for this is that the game engine supports multiple audio formats [...] and are compressed on the fly

That is VERY interesting! I just had a look at Jones3D.exe in Ghidra, I was able to find the WVSM/ADPCM algorithms (very similar to the decompiled code from SWE1TPM, so surprises here) but I couldn't find anything related to the "on the fly compression". I started by searching for "INDYWV" and "WVSM" both in strings and static data, and from there, looked for dependencies. I also searched for the magic values 0x1111 and 0x2222. Am I looking in the wrong place?
But I don't fully understand everything. What is the purpose of compressing WAV data (that is supposedly uncompressed on disk) on the fly? The engine will still need to decompress it later if it wants to play the sounds, right? And I thought that Infernal Machine only had WV files in its game data anyway? (that's the only format I found when I inspected the NDY containers)
Apologies! Maybe I'm missing something here! :)

**Reply by smlu** [reply]
> ..you seem quite familiar with the Sith engine in general

I've been studying the Jones3D engine for quite a while now. Also there is this [OpenJKDF2](https://github.com/shinyquagsire23/OpenJKDF2] project, which tries to re-implement the original sith engine via disassembled code. Really good reference to study the engine internals.

>  I started by searching for "INDYWV" and "WVSM" both in strings and static data, and from there, looked for dependencies. I also searched for the magic values 0x1111 and 0x2222. Am I looking in the wrong place?

If you have access to `Indy3D.exe` search for error message: `Sound_Load: %s is not a WAV or WV file.`
This is the place in sound engine code where sound file header is being parsed.
The code there should look something like this:
```
if ( waveType != 3 && waveType != 6 && waveType != 5 )
{
    Sound_pHS->pErrorPrint("Sound_Load: %s is not a WAV or WV file.\n", filePath);
    Sound_pHS->pFileClose(fh);
    return 0;
}
```

_Of course the variable names will be different in your case._ 

Then a few lines down in if scopes (around 60 lines down) there should be a code similar to this:
```
 if ( Sound_pHS->pFileRead( fh, &soundbank_apSoundCache[pSndInfo->bankNum][dataSize + cacheOffset], *pSndDataSize) != *pSndDataSize )
{
    return 0;
}

// compress data
AudioLib_ResetCompressor(&compressorState);
pSndData_1 = &soundbank_apSoundCache[pSndInfo->bankNum][pSndInfo->dataOffset];
csize = AudioLib_Compress(
            &compressorState,
            pSndData_1 + 4, // out compressed data, 4 bytes seserved for uncompresed size
            &pSndData_1[dataSize],// in uncompressed data
            *pSndDataSize,
            pSndInfo->numChannels);
```

The `AudioLib_Compress` is the function that does compression to `WV` format.
Btw this function is located very close to the one that handles decompression, which you've already found. In my case it's 2 functions above the decompression function.

> But I don't fully understand everything. What is the purpose of compressing WAV data (that is supposedly uncompressed on disk) on the fly? The engine will still need to decompress it later if it wants to play the sounds, right? And I thought that Infernal Machine only had WV files in its game data anyway? (that's the only format I found when I inspected the NDY containers)
Apologies! Maybe I'm missing something here! :)

No worries; you couldn't have known this unless you researched and disassembled the Jones3D engine :)

Here's a little background on the engine:
Originally, the engine was multi-functional; besides being a game engine, it also incorporated a level editor called Marcus. According to interviews with the original developers (mostly Hal Barwood), they encountered enormous problems while developing IM game with this engine. The level editor was removed in the final release, but parts of it remained. Also, my guess is that due to these issues, the engine was compiled in debug release mode, resulting in many assert statements within the engine.

Now, the released engine supports two level file formats. The binary file format, `CND`, contains all level sound data compressed in WV format. The other format is a text-based file format, [NDY](https://github.com/Jones3D-The-Infernal-Engine/Documentation/blob/main/ndy.md), probably a leftover from the level editor. When the engine encounters this file format, audio is loaded on the fly from files on the disk and compressed to WV (if not already in IndyWV format). As for why audio is compressed when loaded from files, my guess is that it was due to limited resources at the time of development, especially on the user-end side. RAM was scarce in 1999, so it was likely more efficient to store audio data in compressed form and decompress it on the fly when needed. Audio compressed in WV format saves approximately half the required space. A good question is why the texture data (MAT) remained uncompressed, although it is about three times less data than the audio data.

**Reply by justedni** [reply]
This is incredible, thank you so much for sharing all this information with me!
And thanks to your guidance, I've managed to find the functions responsible for compressing audio on the fly. Thanks a lot! I'll see if I can analyze the ADPCM compression method when I get a chance :)

And everything you've said about the engine is truly fascinating to me. I'm not surprised at all that they ran into so many problems during development. I'm sure they had very tight deadlines, so the solutions they found seem very clever! Especially with all the technical limitations at the time, like you said, limited RAM, lack of proper support for better audio compression formats, etc.
And I know a lot of games get released with debug info in the final build, even modern ones. But releasing with remnants of the level editor is something else! Very interesting.

**Reply by smlu** [reply]
> Thanks a lot! I'll see if I can analyze the ADPCM compression method when I get a chance :)

Please notify me if you manage to get the compression working with your tool!

Btw if you're gonna work on VW compression algorithm note that there is a buffer overflow bug in the original binary. 
I still haven't wrapped my head around this algorithm and have no idea how exactly works, I just know that if there is clipping in waveform you'll get buffer overflow due to shifted indices in write buffers for left and right channel. You can also ping me when/if you start working on this algorithm and I can try locating those buffer overwrites.

**Reply by justedni** [reply]
I have a working version of the ADPCM compression algo! I've just uploaded the changes to my github (I've left the algorithm in a very raw state for now). I've only done mono ADPCM, because you've already implemented the WVSM algo on your end, right?

I'm having a weird issue though. I've tested Wv -> Wav then Wav -> Wv and Wv -> Wav again, to compare the results. The final Wav plays perfectly (no weird click or anything), but the binaries are slightly different:
![WavToWv_diff](https://github.com/smlu/Urgon/assets/155494991/7cef7b63-af24-418c-a837-e87340ec7054)
I'm not sure if I did something wrong on my end or if it's a bug with the algorithm. I'm guessing the algo should be deterministic!

By the way I wanted to ask you if you had any general advice about a good workflow for this kind of small and localized RE project. I've been using both Ghidra and IDA free, because I prefer the way IDA generates pseudo code. (That said, I've had a bug with WavToWv in IDA. It forgot to add a "multiply by 4" somewhere, but Ghidra had it. it's a bit scary!)
What I do is that I try to rename as many variables as possible in Ghidra and IDA first, but then as soon as I can, I copy/paste to my Visual Studio because it's way easier to iterate and rename everything. When I did WvToWav last month, since this was my first RE project, I spent a LOT of time in x64dbg because there were quite a lot of type errors in the pseudo code: chars replaced by shorts, missing "unsigned", etc. Which was causing an insane amount of runtime errors with this kind of algorithm that does a lot of bitshifting, and/or operators, casting, etc. And maybe I'm having similar issues with WavToWav now.
What's your usual workflow for this? In the code examples you sent me, things like
`Sound_pHS->pFileRead( fh, &soundbank_apSoundCache[pSndInfo->bankNum][dataSize + cacheOffset], *pSndDataSize)`
look horrible in Ghidra:
` (byte *)(**(code **)(DAT_014e4580 + 0x38)) (fh,pbVar5 + iVar8 + (&DAT_014e4530)[piVar7[1]],*pSndDataSize) `
I'm guessing you also move to an external IDE asap, but I'm curious! :)

**Reply by justedni** [reply]
I investigated further, but I didn't find any obvious errors in my RE code. And now I believe that the diffs that I saw in the WVs are there by design! Because if I decompress/compress several times and compare the results (wv->wav, wav->wv, wv->wav and wav->wv one more time), both .wav files are a perfect match even though their .wv sources are slightly different. But better: the 3rd wv is a perfect match with the 2d one! Somehow, those different values produce the same decompressed samples.
I wonder why the 1st .wv file has different values then (I took a random dialog line from SWE1TPM). Is it just because ADPCM is a lossy compression algorithm by design? Meaning that the original .wav file (which we don't have access to) was probably different, but then if we compress/decompress several times then it doesn't destroy the waveform any further?

**Reply by smlu** [reply]
>Is it just because ADPCM is a lossy compression algorithm by design?

That'd be my guess and I strongly believe that is the case. I managed to get uncompressed WAV sound files for IM and comparing it to same audio compressed to WVSM there was significant difference.

>It forgot to add a "multiply by 4" somewhere, but Ghidra had it. it's a bit scary!

Did the assembly had instruction code to multiply by code? Asking this just because there can be also bugs in the original code (e.g.:WVSM compression algo is buggy), tho I had many cases in the past where IDA Hex Ray decompiler forgot to interpret some parts of the assembly, mostly float value comparisons.

>What's your usual workflow for this? In the code examples you sent me, things like
Sound_pHS->pFileRead( fh, &soundbank_apSoundCache[pSndInfo->bankNum][dataSize + cacheOffset], *pSndDataSize)
look horrible in Ghidra:
(byte *)(**(code **)(DAT_014e4580 + 0x38)) (fh,pbVar5 + iVar8 + (&DAT_014e4530)[piVar7[1]],*pSndDataSize)
I'm guessing you also move to an external IDE asap, but I'm curious! :)

TL;DR: I mostly work in RE editor.
Little background story.... I started RE project for IM game with just a goal to get the texture/sound/animation game assets out of CND level files. I did everything in IDA trying to rename variables and functions to some meaningful names reconstruct data structures etc... This project slowly progressed towards more and more code being decompiled. As time passed by I found bits and pieces to references of how the original engine was composed (engine architecture) through reading dev. interviews from the past and talking to some of the OG devs. That sparked the archeology desires within and the urge grew stronger to complete the RE process and unearth the engine code :)
Due to lack of time tho, I managed to extract really tiny portion of the code, just to make some basic engine fixes (injector project similar to OpenJKDF2 project). Working now on extract & cleanup scripts to help me manage the decompiled code to source code files. That being said, I can share parts of code (maybe private repo) with you if you're interested :) just know that the sound part, especial compression/decompression algo is mostly not RE yet (just dumped pseudo code).

**Reply by justedni** [reply]
> Did the assembly had instruction code to multiply by code?

Yes! The instruction is `mov esi, indexes[ebp*4]` but the *4 is nowhere to be found in the generated pseudo code, even earlier in the function. I thought this could be another weird case of incorrect type cast but it's not. Really weird! The lesson I learned here is that I can never trust the generated pseudo code as it is ;)

> Little background story....

Thanks for sharing your RE background with me! :) It's definitely inspiring, and I completely understand you, I know what it is to be passionate about something and to want to push things as far as possible, even if it take years to complete! I'm barely starting on my RE journey but I'm finding this whole process very exciting and rewarding! I'd love to see what you're doing on this private repo of yours! I don't know if I'll understand everything but I'm definitely very curious about your process and current projects :)

**Reply by smlu** [reply]
>Yes! The instruction is mov esi, indexes[ebp*4] but the *4 is nowhere to be found in the generated pseudo code, even earlier in the function. I thought this could be another weird case of incorrect type cast but it's not. Really weird! The lesson I learned here is that I can never trust the generated pseudo code as it is ;)

Oh that is interesting, I think it never occurred to me that the decompiler would left out an instruction like that one. Mostly I'm having issues with float number comparisons which are sometimes skipped by the decompiler. 
Could you maybe paste the code line in your source code where this multiplication occurs?  I'll to cross check with IM code and see if there is similar issue (assuming the code between the games matches).

**Reply by smlu** [reply]
Btw I've recently tested ChatGPT and Claude 3 Sonnet to decompile spinets of assembly code to C pseudo code. The produce quite good struct-ed C code, though you have be curious and double check the outputted code because they can be wrong in certain cases (integer/float comparison for instance). But overall it can help you speed up reverse engineering, especially in cases when decompiler software fails.

**Reply by justedni** [reply]
> Could you maybe paste the code line in your source code where this multiplication occurs?

Sure! In my code it appears in [indywv.cpp](https://github.com/justedni/indywvtowav/blob/main/indywv.cpp?plain=1#L399) on line 399.
If you want to find the source, in my copy of Indy3D.exe, the instructions are around the data offset 004e0108. It looks like that in IDA:
![image](https://github.com/smlu/Urgon/assets/155494991/496d745d-1dc6-44e3-9f31-dd57f0705cea)
The local variable v42 is a __int8, but the data_0054bcf0 is an array of integers, which is _think_ is the reason why v42 should be multiplied by 4 before being used as an index? I don't know why IDA didn't pick it up. But Ghidra does:
![image](https://github.com/smlu/Urgon/assets/155494991/56117d49-8f3e-46ce-8c55-48a6dfd674ea)
And by the way, this is one cool example of RE code that I had to heavily modify in order to make it work in C, after carefully analyzing the static data. The calculated index looks inside an array that contains a list of addresses, that needs to be dereferenced in order to get the actual data! That calculated index can only be 8, 12, 16, 20, 24 or 28 (which is why I've created the "getIndex" function). This was fun to investigate! :)
And one last thing, the problem with the *4 in IDA appears both in SWE1TPM's decompressADPCM, and in Infernal Machine's compressADPCM (I'm pretty sure it's also in IM's decompressADPCM).
But honestly don't spend too much time on this, this could well be a noob mistake on my end! What matters is that it works now :)

**Reply by justedni** [reply]
> Btw I've recently tested ChatGPT and Claude 3 Sonnet to decompile spinets of assembly code to C pseudo code. The produce quite good struct-ed C code, though you have be curious and double check the outputted code because they can be wrong in certain cases (integer/float comparison for instance). But overall it can help you speed up reverse engineering, especially in cases when decompiler software fails.

This is really interesting! I really need to try that. I'm currently looking at another RE project on the side, that involves assembly code from an old MS-DOS game (Cyberia, 1994, Xatrix Entertainment), and I have to say, the 16-bit MS-Dos low-level functions are not well integrated into Ghidra, and IDA stopped handling them a long time ago, so it's making my RE efforts very slow. I've managed to isolate the one procedure that I want to extract (there are a few audio assets in the game are compressed with a proprietary file format called Z94). I'll try putting that into Claude 3 Sonnet! :)

**Reply by smlu** [reply]
I don't think I really every tried to disassemble MS-DOS app before 🤔 I can only guess the support for this platform is veining nowadays. Tho now that I think, there are plenty of dos games I can think which would be fun to peek at it's internals. 

Btw another AI tool I can suggest is mistral (https://chat.mistral.ai). If you gonna try it out it make sure you select `codestral` model which is better tuned/trained for coding.

**Reply by smlu** [reply]
> > Could you maybe paste the code line in your source code where this multiplication occurs?
> 
> Sure! In my code it appears in [indywv.cpp](https://github.com/justedni/indywvtowav/blob/main/indywv.cpp?plain=1#L399) on line 399. If you want to find the source, in my copy of Indy3D.exe, the instructions are around the data offset 004e0108. It looks like that in IDA: ![image](https://private-user-images.githubusercontent.com/155494991/342812222-496d745d-1dc6-44e3-9f31-dd57f0705cea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzMDc1NjYsIm5iZiI6MTcyMTMwNzI2NiwicGF0aCI6Ii8xNTU0OTQ5OTEvMzQyODEyMjIyLTQ5NmQ3NDVkLTFkYzYtNDRlMy05ZjMxLWRkNTdmMDcwNWNlYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOFQxMjU0MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOTliNWUxYmIwNDQ0ZDQ5NmRhYzlhZGY0YzAzMTUwYTZiYTdiYTdhM2JhMjE1MGIxNzQwMzZhNjA4OTgxY2VkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.UdwJ7pbAHce-x8jnZFvkmdNOPJQ_wa8-u9bRspIyoiE) The local variable v42 is a __int8, but the data_0054bcf0 is an array of integers, which is _think_ is the reason why v42 should be multiplied by 4 before being used as an index? I don't know why IDA didn't pick it up. But Ghidra does: ![image](https://private-user-images.githubusercontent.com/155494991/342813868-56117d49-8f3e-46ce-8c55-48a6dfd674ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzMDc1NjYsIm5iZiI6MTcyMTMwNzI2NiwicGF0aCI6Ii8xNTU0OTQ5OTEvMzQyODEzODY4LTU2MTE3ZDQ5LThmM2UtNDZjZS04YzU1LTQ4YTZkZmQ2NzRlYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOFQxMjU0MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMDdkNzk4NTA4N2RlOWY3NDYzMzc5NTdkOTI0OTc0M2JmZDNhZTdhMDhlYTdkZGRlNzQxMTFmNmRmMzBmODhmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.YN3Uaxfzumgq-XKBFHrTFjUBwi-NEqHqfVWwQt0OoUs) And by the way, this is one cool example of RE code that I had to heavily modify in order to make it work in C, after carefully analyzing the static data. The calculated index looks inside an array that contains a list of addresses, that needs to be dereferenced in order to get the actual data! That calculated index can only be 8, 12, 16, 20, 24 or 28 (which is why I've created the "getIndex" function). This was fun to investigate! :) And one last thing, the problem with the *4 in IDA appears both in SWE1TPM's decompressADPCM, and in Infernal Machine's compressADPCM (I'm pretty sure it's also in IM's decompressADPCM). But honestly don't spend too much time on this, this could well be a noob mistake on my end! What matters is that it works now :)



Looking at my de-compiled code I have this:
```C
v36 = v22;
*(_WORD *)v29 = v22;
v29 += sampleSize;
v24 = AudioLib_aIndexTableTable[stepBits][v18] + a2a;
if ( v24 > 0 ) {
  if ( v24 >= 88 )
      v24 = 88;
}
else {
  v24 = 0;
}
```

The `AudioLib_aIndexTableTable` should be `data_0054bcf0` in your case, I think.
And I think you're right that the array contains the addresses to other arrays.

This is the definition of  `AudioLib_aIndexTableTable` :
```C
const int8_t *AudioLib_aIndexTableTable[8] =
{
  NULL,
  NULL,
  &AudioLib_aIndex2Bit,
  &AudioLib_aIndex3Bit,
  &AudioLib_aIndex4Bit,
  &AudioLib_aIndex5Bit,
  &AudioLib_aIndex6Bit,
  &AudioLib_aIndex7Bit
};
```
The `AudioLib_aIndex*` are arrays of different dimensions.

In my case I think the decompiler got it right there by accessing the `AudioLib_aIndexTableTable` like it's 2D array.
Btw the I just remembered that the symbol names (e.g.: `AudioLib_aIndexTableTable`) I got from linux version of Grim Fandango Remastered. Fyi it contains all debugging symbols 😃 

Also recently I discovered debug version of Indy3D.exe on archive.org
Here's the link: https://archive.org/download/2-indiana-jones/2%20Indiana%20Jones.rar/La%20Macchina%20Infernale%2FResource%2FCD1.GOB  
You have to extract content of gob, and the larger one (~2MB) of the 2 Indy3D.exe is debug version.
What's great about this version is that due to being combined in debug mode the code is not optimized and decompiled code could better resemble the original code in some cases.

**Reply by justedni** [reply]
Sorry for the late reply! Thanks a lot for looking into it! It makes perfect sense to me now, I think the problem came from my re-implementation of AudioLib_aIndexTableTable ([indywv_data.cpp:47](https://github.com/justedni/indywvtowav/blob/main/indywv_data.cpp?plain=1#L47)) I was using both Ghidra and IDA to re-implement this and I probably got confused at some point. The "raw" version of AudioLib_aIndexTableTable contained 4-bytes addresses, which is why the index parameter to access the array (stepBits in your case) needed to be multiplied by 4 in order to point to the wanted address in the raw array. But when it's represented as a 2D array like in your case, then the code shouldn't multiply by 4. Thanks for your help!

And I had no idea that there were versions of those games with debug symbols and everything! That's an amazing find!! I need to check it out!! :)

By the way, thanks again for suggesting Claude AI, I've been using it on a regular basis since you mentioned it, it's great for quick prototyping (C++ or python in my case), and it also looks very promising for RE. I sent it raw MS-Dos assembly files and it was able to analyse and decompile portions of the code. The MS-Dos RE project I'm working on atm is a bit complex but it definitely speeds things up, having an AI giving quick insights on some obscure functions that I'd have to google one-by-one otherwise. I'll check out Mistral as well one of those days!

**Reply by smlu** [reply]
> And I had no idea that there were versions of those games with debug symbols and everything! That's an amazing find!! I need to check it out!! :)

Unfortunately in the case of Indy3D.exe the pdb file is missing but the function structure resembles the original code (e.g.: preserved functions which should be inlined). But the remastered linux version of Grim Fandango is I guess is gold digger in terms of debug symbols :)
